src/bounded/sat.rs

```diff
@@ -0,0 +1,811 @@
+            return Err(CheckerError::UnknownSort(sort.clone(), universe.clone()));
+        }
+    }
+
+    for relation in &module.signature.relations {
+        if relation.sort != Sort::Bool {
+            todo!("non-bool relations")
+        }
+    }
+
+    if !module.defs.is_empty() {
+        todo!("definitions are not supported yet");
+    }
+
+    let mut assumes = Vec::new();
+    let mut asserts = Vec::new();
+    for statement in &module.statements {
+        match statement {
+            ThmStmt::Assert(Proof { assert, invariants }) => {
+                asserts.push(assert.x.clone());
+                if !invariants.is_empty() {
+                    eprintln!("note: invariants are not yet supported, and do nothing")
+                }
+            }
+            ThmStmt::Assume(term) if asserts.is_empty() => assumes.push(term.clone()),
+            _ => return Err(CheckerError::OutOfOrderStatement(statement.clone())),
+        }
+    }
+
+    let mut inits = Vec::new();
+    let mut trs = Vec::new();
+    for assume in assumes {
+        match assume {
+            Term::UnaryOp(UOp::Always, axiom) if FirstOrder::unrolling(&axiom) == Some(0) => {
+                inits.push(*axiom.clone());
+                trs.push(*axiom);
+            }
+            Term::UnaryOp(UOp::Always, tr) if FirstOrder::unrolling(&tr) == Some(1) => {
+                trs.push(*tr)
+            }
+            Term::UnaryOp(UOp::Always, term) => return Err(CheckerError::TooFuture(*term)),
+            init if FirstOrder::unrolling(&init) == Some(0) => inits.push(init),
+            init => return Err(CheckerError::AnyFuture(init)),
+        }
+    }
+
+    let mut safes = Vec::new();
+    for assert in asserts {
+        match assert {
+            Term::UnaryOp(UOp::Always, safe) if FirstOrder::unrolling(&safe) == Some(0) => {
+                safes.push(*safe)
+            }
+            Term::UnaryOp(UOp::Always, safe) => return Err(CheckerError::AnyFuture(*safe)),
+            assert => return Err(CheckerError::AssertWithoutAlways(assert)),
+        }
+    }
+
+    let mut context = Context::new(&module.signature, universe, depth);
+
+    let translate = |terms| {
+        let term = Term::NAryOp(NOp::And, terms);
+        let term = nullary_id_to_app(term, &module.signature.relations);
+        let term = crate::term::Next::new(&module.signature).normalize(&term);
+        term_to_ast(&term, &context, &HashMap::new())
+    };
+
+    println!("starting translation...");
+    let translation = std::time::Instant::now();
+
+    let init = translate(inits)?;
+    let tr = translate(trs)?;
+    let not_safe = Ast::Not(Box::new(translate(safes)?));
+
+    let mut program = vec![init];
+    for i in 0..depth {
+        program.push(tr.clone().prime(i));
+    }
+    program.push(not_safe.prime(depth));
+
+    let cnf = tseytin(&Ast::And(program), &mut context);
+
+    println!("translation finished in {:?}", translation.elapsed());
```

Comment:
Is there much point in printing this? Won't it basically always be fast?

Reply:
It should be fast, but a) it brings it in line with the other checkers, where translation is slow, and b) if it becomes slow we really want to know.